components/uuid-generator/org.wso2.carbon.uuid.generator/src/main/java/org/wso2/carbon/uuid/generator/TimeBasedUUIDGenerator.java

```diff
@@ -0,0 +1,108 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.uuid.generator;
+
+import org.wso2.carbon.uuid.generator.util.UUIDGeneratorConstants;
+import org.wso2.carbon.uuid.generator.util.Util;
+
+import java.security.SecureRandom;
+import java.util.Random;
+import java.util.UUID;
+
+public class TimeBasedUUIDGenerator extends UUIDGenerator{
+
+    long _address;
+    //Last 8 bytes of the UUID (node id + clock sequence)
+    protected final long uuidSecondHalf;
+
+    protected final UUIDTimeStamp uuidTimeStamp;
+
+    public TimeBasedUUIDGenerator(UUIDTimeStamp timeStamp) {
+
+        byte[] timeBaseduuidByteAray = new byte[16];
+        constructNodeId();
+        insertToByteArray(timeBaseduuidByteAray, 10);
+        int clockSequence = timeStamp.getClockSequence();
+        //set clock_seq_hi.The high field of the clock sequence.
+        timeBaseduuidByteAray[UUIDGeneratorConstants.CLOCK_SEQUENCE_HI_BYTE_OFFSET_IN_UUID] =
+                (byte) (clockSequence >> 8);
+        //set clock_seq_low.
+        timeBaseduuidByteAray[UUIDGeneratorConstants.CLOCK_SEQUENCE_LOW_BYTE_OFFSET_IN_UUID] = (byte) clockSequence;
+        long clockSeqInLong = Util.createClockSeqInLong(timeBaseduuidByteAray, 8);
+        uuidSecondHalf = Util.createUUIDSecondHalfInLong(clockSeqInLong);
+        uuidTimeStamp = timeStamp;
+    }
+
+    @Override
+    public UUID generate()
+    {
+        final long rawTimestamp = uuidTimeStamp.getTimestamp();
+        // Time field components are kind of shuffled, need to slice:
+        int clockHi = (int) (rawTimestamp >>> 32);
+        int clockLo = (int) rawTimestamp;
+        // and dice
+        int midhi = (clockHi << 16) | (clockHi >>> 16);
+        // need to squeeze in type (4 MSBs in byte 6, clock hi)
+        midhi &= ~0xF000; // remove high nibble of 6th byte
+        midhi |= 0x1000; // type 1
+        long midhiL = (long) midhi;
+        midhiL = ((midhiL << 32) >>> 32); // to get rid of sign extension
+        // and reconstruct
+        long l1 = (((long) clockLo) << 32) | midhiL;
+        // last detail: must force 2 MSB to be '10'
+        return new UUID(l1, uuidSecondHalf);
+    }
+
+    private byte[] constructNodeId(){
+
+        Random random = new SecureRandom();
+        byte[] nodeIdByteArray = new byte[6];
+        synchronized (random) {
+            random.nextBytes(nodeIdByteArray);
+        }
+        nodeIdByteArray[0] |= (byte) 0x01;
+        createBinaryAddress(nodeIdByteArray);
+        return nodeIdByteArray;
+    }
+
+    private void createBinaryAddress(byte[] nodeIdByteArray){
+        if (nodeIdByteArray.length != 6) {
+            throw new NumberFormatException("Ethernet address has to consist of 6 bytes");
```

Comment:
What's the relation with the ethernet address

Reply:
This should be node id address.
According to RFC this is the description for Node id

 For UUID version 1, the node field consists of an IEEE 802 MAC
   address, usually the host address.  For systems with multiple IEEE
   802 addresses, any available one can be used.  The lowest addressed
   octet (octet number 10) contains the global/local bit and the
   unicast/multicast bit, and is the first octet of the address
   transmitted on an 802.3 LAN.

   For systems with no IEEE address, a randomly or pseudo-randomly
   generated value may be used; see Section 4.5.  The multicast bit must
   be set in such addresses, in order that they will never conflict with
   addresses obtained from network cards.

Reply:
Fixed with 05f8b08